keps/sig-apps/3939-consider-terminating-pods-jobs/README.md

```diff
@@ -0,0 +1,712 @@
+The Job controller does not typically wait for terminating pods to be marked as failed.  
+Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
+See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
+Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+One area of contention is how this KEP will work with 3329-retriable-and-non-retriable-failures.  It is important to mention the subtleties here.  
+
+In 3329, there was a decision to force kubelet to transition pods to failed before marking them as failed.  This is feature toggled guarded by `PodDisruptionCondition`.  
+
+This means that when this feature is turned on we will only mark pods as failed once they are fully terminated.  This means that the pod is still considered active.  
+
+If this feature is turned off then a pod is marked as failed as soon as it is terminated so it will be counted as failed rather than active.  
+
+To get around this problem we decided to add a field called terminating so we can count separately from active or failed.  
+
+Another issue is described here: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509.  
+
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase. As a consequence, it will never be replaced, so the whole job will be stuck from making progress.
+
+Due to the above issue, we can set phase=Failed when PodDisruptionConditions is enabled OR JobRecreatePodsWhenFailed is enabled. When JobRecreatePodsWhenFailed enabled, but PodDisruptionConditions disabled we would just set the phase, but without adding the condition.
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new enum field:
+
+```golang
+// This field controls when we recreate pods
+// Default will be TerminatingOrFailed ie recreate pods when they are failed
+// +enum 
+type RecreatePodsWhen string
+const (
+ // This is a field that recreates pods when they are marked as terminating or failed
+ // ie this means that as soon as pods get marked for deletion they will be recreated
+ TerminatingOrFailed RecreatePodsWhen = "TerminatingOrFailed"
+ // Only recreate pods when they are marked as failed.
+ Failed              RecreatePodsWhen = "Failed"
```

Comment:
`RecreatePodsWhen` field could work for Deployments as well. The values probably not. We should at least make sure that we sync between Job and Deployment when adding new values to prevent same concepts named differently.

We should make sure the synonyms and shape is consistent. For example we might be adding the following values to deployments: `TerminatingOrTerminal`, `Terminal`

I wonder if there should be some process about these naming conventions? Should we drop a comment here and in deployments about this?

Reply:
The reason we split this as it still seems that the API for Deployment is not set yet.  I'm happy for some help on the deployment side.  I think we decided to split partially because we think the Job KEP is ready for review and there is still a lot of discussion on the deployment API.  And it seems that there is not likely going to be a similar API for both at this moment.  Other reason was so that we fully decouple these two KEPS from each other.  

I think the implementation is similar but there seems to be some more dicussion needed for the deployment side.

Reply:
I agree, it is probably better to focus on the Job first. We should just keep in back of our mind that it could be relevant for deployments, that is all. 

Ping me once you switch the attention to deployments. In case you will not have a time to focus on the other KEP, I can take over if you want.

Reply:
I'll let you know.  I will not be able to get to Deployments this release and I'll see how the code review goes for this KEP.